pallets/api/src/messaging/mod.rs

```diff
@@ -386,174 +596,459 @@ pub mod pallet {
-					id,
-					callback,
-				});
-				Self::deposit_event(Event::<T>::Removed {
-					origin: origin.clone(),
-					messages: [id].to_vec(),
-				});
 				Ok(())
-			},
-			Err(error) => {
-				// Fallback to storing the message for polling - pre-paid weight is lost.
-				Self::deposit_event(Event::<T>::CallbackFailed {
-					origin: origin.clone(),
-					id,
-					callback,
-					error: error.error,
-				});
-				// TODO: logging
-				Err(error.error)
-			},
+			})?;
+
+			Ok(())
 		}
 	}
-}
 
-#[derive(Clone, Decode, Debug, Encode, MaxEncodedLen, PartialEq, TypeInfo)]
-pub enum Status {
-	Pending,
-	TimedOut,
-	Complete,
-}
+	impl<T: Config> Pallet<T> {
+		/// Executes a registered callback with the given input data and manually charges block
+		/// weight.
+		///
+		/// This function is responsible for handling the full lifecycle of a callback invocation:
+		/// - Calculating the total weight cost of the callback, including a static adjustment.
+		/// - Ensuring that sufficient blockspace is available before execution.
+		/// - Executing the callback via the configured `CallbackExecutor`.
+		/// - Registering the actual weight used with the runtime.
+		/// - Managing any associated weight fee refund logic.
+		///
+		/// # Parameters
+		///
+		/// - `initiating_origin`: The account that triggered the callback. This will be passed to
+		///   the executor and used for fee management and event attribution.
+		/// - `callback`: The callback definition.
+		/// - `id`: The message ID associated with this callback's message.
+		/// - `data`: The encoded payload to send to the callback.
+		/// - `static_weight_adjustment`: An optional additional weight to charge, typically to
+		///   account for outer logic like wrapping or delegation. This is added to the callback's
+		///   declared weight.
+		///
+		/// # Weight Handling
+		///
+		/// - Before executing the callback, this function checks whether the total expected weight
+		///   (`callback.weight + static_weight_adjustment`) can be accommodated in the current
+		///   block.
+		/// - If the block is saturated, the function returns early with an error and does not
+		///   mutate state.
+		/// - After execution, the actual weight used by the callback is determined using
+		///   [`Self::process_callback_weight`] and registered via
+		///   [`frame_system::Pallet::<T>::register_extra_weight_unchecked`].
+		pub(crate) fn call(
+			initiating_origin: &AccountIdOf<T>,
+			callback: Callback,
+			id: &MessageId,
+			data: &impl Encode,
+			// This should include T::CallbackExecutor::execution_weight() where applicable unless
+			// already charged.
+			static_weight_adjustment: Option<Weight>,
+		) -> DispatchResult {
+			// This is the total weight that should be deducted from the blockspace for callback
+			// execution.
+			let max_weight = callback
+				.weight
+				.checked_add(&static_weight_adjustment.unwrap_or(Zero::zero()))
+				.ok_or(Error::<T>::BlockspaceAllowanceReached)?;
```

Comment:
Potentially we need to return a weight from call. 
If call is an error the return Ok(weight) and wrap it in a call_and_mutate_blockweight function. 

In the current state we are not chaeging any blockspace if a call fails using a static weight adjustment > 0. Which could possibly be exploitied.